### PR TITLE
[MODULAR]Modular Disabler PR - UNFINISHED. [WIP][DRAFT]

### PR DESCRIPTION
**Pull request by nevimer**
## About The Pull Request

Right! So, this is still very much a work in progress. I'm mostly opening a PR for this to showcase what I am working on regarding these. 

Normal disablers can be upgraded into a modular disabler with an upgrade kit available by completing toxins first test. 

As a tradeoff to the upgrade, the time to stun is significantly lower, but have self recharging ammunition (like a PKA).

The upgraded disablers expel water and sparks when firing, further reducing their effectiveness in fights due to the slip risk and heat they produce. Right now it has the backblast, but I think that goes too far and is abusable. **This is going to be removed cause it sucks.** 

Also. This is probably awful code but do tell how to improve!

Have some videos from testing.
## Pushing shots:
https://user-images.githubusercontent.com/77420409/154829164-609f7db1-81cd-49fa-bef3-c9f21762b9a0.mp4

## Hallucination (Subject to Change)
https://user-images.githubusercontent.com/77420409/154829208-0c5514d4-ccd2-43f1-b901-d5660c8f51cf.mp4

## Applying pushing shot upgrade to an upgrade disabler

https://user-images.githubusercontent.com/77420409/154829415-a1bfb38d-c73a-4900-b8f4-da462d088b09.mp4

## WHAT?

Modular disablers! You attach things to them and make your disabler do things! Currently: New shot modifiers, and infinite ammo!

## WHERE?

Antag items and security items! Right after toxins, hopefully.

## WHO?

Hopefully these are in lots of peoples hands!

## WHY?

I was bored and wanted to practice writing things. It's clear that we the players are currently unhappy with the current non lethal options, and are far too lethal. This should move us away from one hit stuns and lethals as the solution to PvP encounters. 

## How This Contributes To The Skyrat Roleplay Experience

HOPEFULLY this makes both antags and security players have a bit more fun in their combat gameplayer, with weapons that are *not* lethal or stun based. 

## Quit touching security!!

Things could be better! This I hope reduces reliance on stun based combat.

## I hate this, I am going to leave a thumbs down and complain about it
Okay then. 

## Changelog

:cl:
add: Nanotrasen Research and Development has released a new line of disabler barrel attachments. Disablers can now be upgraded to be modifiable, at a cost. 
code: Moves all the skyrat edits into the modular file, restores the TG versions where applicable. 

/:cl:

### CODE REVIEW

File: modular_skyrat/modules/modular_disabler/beams.dm [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+/obj/projectile/beam/disabler/weak/split
+	damage = 12
+
+/obj/projectile/beam/disabler/weak/untie
+	damage = 15
+	light_color = COLOR_SAMPLE_BROWN
+
+/obj/projectile/beam/disabler/weak/untie/on_hit(atom/target, blocked, pierce_hit)
+	. = ..()
+	var/mob/living/carbon/hit = target
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/beams.dm:14]: you could typecast the `target` argument to mob/living/carbon and just do an early return if an `istype(target)` fails

File: modular_skyrat/modules/modular_disabler/beams.dm [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	if(iscarbon(hit) && prob(75))
+		shoe.adjust_laces(SHOES_UNTIED, src)
+
+
+/obj/projectile/beam/disabler/weak/shove
+	damage = 18
+	projectile_piercing = PASSMOB
+/obj/projectile/beam/disabler/weak/shove/on_hit(atom/target, blocked, pierce_hit)
+	. = ..()
+	var/mob/living/carbon/hit = target
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/beams.dm:25]: ditto

File: modular_skyrat/modules/modular_disabler/beams.dm [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		buckle_mob(hit, TRUE, FALSE)
+
+/obj/projectile/beam/disabler/disgust // Disgusting beam...
+	damage = 0
+	icon_state = "omnilaser"
+	light_color = LIGHT_COLOR_GREEN
+
+/obj/projectile/beam/disabler/disgust/on_hit(atom/target, blocked, pierce_hit)
+	. = ..()
+	var/mob/living/carbon/hit = target
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/beams.dm:36]: ditto

File: modular_skyrat/modules/modular_disabler/beams.dm [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+/obj/projectile/beam/disabler/warcrime
+	damage = 0
+	icon_state = "toxin"
+	light_color = LIGHT_COLOR_BLOOD_MAGIC
+
+/obj/projectile/beam/disabler/warcrime/on_hit(atom/target, blocked, pierce_hit) // Might be a Traitor(sec) thing.
+	. = ..()
+	var/mob/living/carbon/human/hit = target
+	var/effects = rand(5,15)
+	if(ishuman(hit))
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/beams.dm:49]: ditto

File: modular_skyrat/modules/modular_disabler/beams.dm [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+/obj/projectile/beam/disabler/drowsy
+	icon_state = "omnilaser"
+	light_color = LIGHT_COLOR_LAVENDER
+
+/obj/projectile/beam/disabler/drowsy/on_hit(atom/target, blocked, pierce_hit)
+	. = ..()
+	var/mob/living/carbon/hit = target
+	if(iscarbon(hit))
+		hit.adjust_drowsyness(rand(10,20))
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/beams.dm:66]: ditto

File: modular_skyrat/modules/modular_disabler/beams.dm [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+/obj/projectile/beam/disabler/hallucinate // Make an officer chase the suspect
+	damage = 0
+	icon_state = "omnilaser"
+	light_color = LIGHT_COLOR_GREEN
+
+/obj/projectile/beam/disabler/hallucinate/on_hit(atom/target, blocked, pierce_hit)
+	. = ..()
+	var/mob/living/carbon/hit = target
+	var/datum/brain_trauma/special/beepsky/beepsky_hallucination
+	if(iscarbon(hit))
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/beams.dm:77]: ditto and early return

File: modular_skyrat/modules/modular_disabler/beams.dm [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+		if(beepsky_hallucination)
+			return
+		beepsky_hallucination = new()
+		hit.gain_trauma(beepsky_hallucination, TRAUMA_RESILIENCE_ABSOLUTE)
+		spawn(30 SECONDS)
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/beams.dm]: don't solely take my word for it but I'm pretty sure everything that needs to be delayed after the `spawn()` needs to be indented

File: modular_skyrat/modules/modular_disabler/disabler_modular.dm [hunk 8]
```diff
@@ -0,0 +1,90 @@
+/obj/item/gun/energy/disabler
+	cell_type = /obj/item/stock_parts/cell/super
+
+
+/obj/item/gun/energy/disabler/upgraded/proc/addammotype(ammo_to_add, name_to_append)
+	ammo_type += new ammo_to_add(src)
+	name += " of [name_to_append]"
+/obj/item/gun/energy/disabler/upgraded
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/disabler_modular.dm:8]: code doc comments

File: modular_skyrat/modules/modular_disabler/disabler_modular.dm [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+	var/max_mod_capacity = 100
+	var/list/modkits = list()
+	var/upgrade_name
+	var/amount = 0
+	var/maxamount = 3
+	var/emission_amount = 1
+	var/recharge_timerid
+	var/backblast = TRUE
+	name = "upgraded disabler"
+	Initialize(mapload)
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/disabler_modular.dm:26]: never use relative pathing

File: modular_skyrat/modules/modular_disabler/disabler_modular.dm [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+	var/max_mod_capacity = 100
+	var/list/modkits = list()
+	var/upgrade_name
+	var/amount = 0
+	var/maxamount = 3
+	var/emission_amount = 1
+	var/recharge_timerid
+	var/backblast = TRUE
+	name = "upgraded disabler"
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/disabler_modular.dm:25]: name should be at the very top, above all the custom variables

File: modular_skyrat/modules/modular_disabler/disabler_modular.dm (hunk 9 above)
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/disabler_modular.dm:26]: add a line break between the vars and the procs

File: modular_skyrat/modules/modular_disabler/disabler_modular.dm [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+		else
+			carried = 1
+
+		deltimer(recharge_timerid)
+		recharge_timerid = addtimer(CALLBACK(src, .proc/reload), recharge_time * carried, TIMER_STOPPABLE)
+
+	emp_act(severity)
+		return
+	proc/emissions(recharge_time, emission_amount)
+		var/turf/open/T = get_turf(src)
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/disabler_modular.dm:69]: one letter vars are bad

File: modular_skyrat/modules/modular_disabler/disabler_modular.dm [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+			return
+		if(overheat)
+			return
+		if(!recharge_time)
+			recharge_time = overheat_time
+		overheat = TRUE
+		emissions(recharge_time, emission_amount)
+		var/carried = 0
+		if(!unique_frequency)
+			for(var/obj/item/gun/energy/disabler/upgraded/K in loc.get_all_contents())
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/disabler_modular.dm:55]: one letter vars are bad

File: modular_skyrat/modules/modular_disabler/kits.dm [hunk 13]
```diff
@@ -0,0 +1,19 @@
+/obj/item/device/custom_kit/disabler
+	var/ammo_to_add
+	var/name_to_append
+
+/obj/item/device/custom_kit/disabler/afterattack(obj/target, mob/user, proximity_flag)
+	var/obj/item/gun/energy/disabler/upgraded/target_obj = target
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/kits.dm:6]: typecast the `target` argument to `/obj/item/gun/energy/disabler/upgraded`, do an `istype(target)` check here with an early return instead

File: modular_skyrat/modules/modular_disabler/kits.dm [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+	if(!proximity_flag) //Gotta be adjacent to your target
+		return
+	if(isturf(target_obj)) //This shouldn't be needed, but apparently it throws runtimes otherwise.
+		return
+	if(target_obj.amount >= target_obj.maxamount)
+		to_chat(user, span_warning("You can't improve [target_obj] any further!"))
+		return
+	target_obj.addammotype(ammo_to_add, _name_to_append)
+	user.visible_message(span_notice("[user] modifies [target_obj] with [ammo_to_add]!"), span_notice("You modify [target_obj] with [ammo_to_add]."))
+	qdel(src)
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/kits.dm:17]:
```suggestion
	QDEL_NULL(src)
```

File: modular_skyrat/modules/modular_disabler/upgraded_templates.dm [hunk 15]
```diff
@@ -0,0 +1,14 @@
+/obj/item/gun/energy/disabler/upgraded/disgust // Disgusting Disabler...
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/upgraded_templates.dm:1]: Why is this here?

File: modular_skyrat/modules/modular_disabler/upgraded_templates.dm [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+/obj/item/gun/energy/disabler/upgraded/disgust/Initialize(mapload)
+	. = ..()
+	var/ammo_to_load = /obj/item/ammo_casing/energy/disabler/skyrat/proto/disgust
+	ammo_type += new ammo_to_load(src)
+	name += " of [SHOT_DISGUST]"
+
+
+/obj/item/gun/energy/disabler/upgraded/warcrime // Antag Item Template (Intended to buy this with TC)
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/upgraded_templates.dm:9]: same goes for here

File: modular_skyrat/modules/modular_disabler/uplink_items.dm [hunk 17]
```diff
@@ -0,0 +1,14 @@
+/datum/uplink_item/dangerous/disabler_modular
+	name = "Prototype Modular Disabler"
+	desc = "An exceedingly rare prototype modular disabler, fitted with an \
+	illusive barrel marked \'Warcrimes\'. Fancy! Security wishes they had these."
+	item = /obj/item/gun/energy/disabler/upgraded/warcrime
+	cost = 6
+	surplus = 2
+/datum/uplink_item/ammo/warcrime_disabler
+	name = "Warcrime Barrel"
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/uplink_items.dm:9]: ideally a more consise name

File: modular_skyrat/modules/modular_disabler/uplink_items.dm [hunk 18]
```diff
@@ -0,0 +1,14 @@
+/datum/uplink_item/dangerous/disabler_modular
+	name = "Prototype Modular Disabler"
+	desc = "An exceedingly rare prototype modular disabler, fitted with an \
+	illusive barrel marked \'Warcrimes\'. Fancy! Security wishes they had these."
+	item = /obj/item/gun/energy/disabler/upgraded/warcrime
+	cost = 6
+	surplus = 2
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_disabler/uplink_items.dm:7]: note that this knocks down the chance of it being in a surplus crate to 2%

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by KathrinBailey** [comment]
Wouldn't the backblast just set you on fire? It's not a shoulder-mounted disabler or something.
Also, the water is very meme-y, why would we want slip combat to ever rear its ugly head like this? Half the server were sick and tired of it back in the Cit days.

**Comment by nevimer** [comment]
> Wouldn't the backblast just set you on fire? It's not a shoulder-mounted disabler or something.
> Also, the water is very meme-y, why would we want slip combat to ever rear its ugly head like this? Half the server were sick and tired of it back in the Cit days.

Water emissions and backblast are going away. They do suck. Water is probably going to be replaced with a different gas, maybe miasma? or just co2.

**Comment by nevimer** [comment]
Honestly, fuck this. If people are just going to downvote it while it is a draft and not even nearing done, I'll just concentrate my efforts some where else.

**Comment by nevimer** [comment]
New name, since I probably will be removing a bunch of things and don't want to make promises in a title.

**Comment by Zergspower** [comment]
holy shit that's a huge buff...

**Comment by ArdentSpark** [comment]
To be fair, it's not inherantly a buff. It's an upgrade system. It has the potential to be imbalanced, but balancing is easier than coding this system from scratch, and can always come after it's finished development if required

> Normal disablers can be upgraded into a modular disabler with an upgrade kit available by completing toxins first test.
> 
> As a tradeoff to the upgrade, the time to stun is significantly lower, but have self recharging ammunition (like a PKA).

Although I do think this is supposed to say "Higher", as lower time to stun means quicker stuns.
After discussing with Nevimer, we kind of came to the conclusion that the MCR is one of the most fun things in the security arsenal. If the balance is done right with these I don't think it's a bad idea, even if I disagree with limitless disablers

(also this is a draft, I'm not sure we should be worrying too much about balance at this stage.)

**Comment by Zergspower** [comment]
> To be fair, it's not inherantly a buff. It's an upgrade system. It has the potential to be imbalanced, but balancing is easier than coding this system from scratch, and can always come after it's finished development if required
> 
> > Normal disablers can be upgraded into a modular disabler with an upgrade kit available by completing toxins first test.
> > As a tradeoff to the upgrade, the time to stun is significantly lower, but have self recharging ammunition (like a PKA).
> 
> Although I do think this is supposed to say "Higher", as lower time to stun means quicker stuns. After discussing with Nevimer, we kind of came to the conclusion that the MCR is one of the most fun things in the security arsenal. If the balance is done right with these I don't think it's a bad idea, even if I disagree with limitless disablers
> 
> (also this is a draft, I'm not sure we should be worrying too much about balance at this stage.)

i'm talking about the yeet them 6 tiles after hitting them once mostly - and definitely the self-recharge is too much imo. MCR's may be fun but they are currently overpowered as hell as well :<

**Comment by nevimer** [comment]
On hold for now.

**Pull request closed without merging**